### PR TITLE
Enhance Metadata Filtering with Parent ID and Label matching support

### PR DESCRIPTION
**Pull request by jackbackes**
Enhanced the ListContent API to support advanced filtering options. This extension allows users to filter content based on multiple criteria such as `parent_id`, and a new `labels_eq_filter`. The update includes:

- Added `parent_id` and `labels_eq_filter` fields in `ListContentRequest`.
- Implemented new utility functions in `api_utils.rs` for label validation and deserialization.
- Updated `ListContentParams` struct to include `parent_id` and `labels_eq`.
- Created `coordinator_filters.rs` for filtering logic in the content listing.
- Modified `CoordinatorService` to utilize new filters in the `list_content` method.
- Adjusted `DataRepositoryManager` and `server.rs` to accommodate new filter parameters.
- Added to Getting Started docs

This update significantly improves the flexibility of the content querying process, allowing for more precise and efficient data retrieval.

Tests:
- All unit tests and integration tests pass (`cargo test -- --test-threads 1`).
- Thoroughly validated new filters and utility functions for robustness and edge cases.

### CODE REVIEW

File: src/data_repository_manager.rs [hunk 1]
```diff
@@ -216,10 +216,18 @@ impl DataRepositoryManager {
...
         repository: &str,
         source_filter: &str,
+        parent_id_filter: &str,
+        labels_eq: Option<&HashMap<String, String>>,
     ) -> Result<Vec<api::ContentMetadata>> {
         let req = indexify_coordinator::ListContentRequest {
             repository: repository.to_string(),
             source: source_filter.to_string(),
+            parent_id: parent_id_filter.to_string(),
+            has_labels_eq_filter: labels_eq.is_some(),
```
Comment (diptanu) [inline src/data_repository_manager.rs:226]: Can we remove this?

File: src/data_repository_manager.rs [hunk 2]
```diff
@@ -216,10 +216,18 @@ impl DataRepositoryManager {
...
+            labels_eq_filter: match labels_eq {
```
Comment (diptanu) [inline src/data_repository_manager.rs:227]: labels_eq.unwrap_or(HashMap::new())

File: src/indexify_coordinator.rs [hunk 3]
```diff
@@ -216,6 +216,15 @@ pub struct ListContentRequest {
     pub repository: ::prost::alloc::string::String,
     #[prost(string, tag = "2")]
     pub source: ::prost::alloc::string::String,
+    #[prost(string, tag = "3")]
+    pub parent_id: ::prost::alloc::string::String,
+    #[prost(bool, tag = "4")]
+    pub has_labels_eq_filter: bool,
```
Comment (diptanu) [inline src/indexify_coordinator.rs:222]: No need for this

File: src/coordinator_service.rs [hunk 4]
```diff
@@ -100,9 +100,13 @@ impl CoordinatorService for CoordinatorServiceServer {
         request: tonic::Request<ListContentRequest>,
     ) -> Result<tonic::Response<ListContentResponse>, tonic::Status> {
         let req = request.into_inner();
+        let labels_eq = match req.has_labels_eq_filter {
```
Comment (diptanu) [inline src/coordinator_service.rs:103]: Just pass in empty HashMap?

**Review by diptanu — CHANGES_REQUESTED** [review]

File: src/api.rs [hunk 5]
```diff
@@ -278,9 +279,20 @@ impl From<attribute_index::ExtractedMetadata> for ExtractedMetadata {
     }
 }
 
-#[derive(Debug, Serialize, Deserialize, ToSchema)]
-pub struct ContentSourceFilter {
+#[derive(Debug, Serialize, Deserialize, ToSchema, PartialEq, Clone)]
+pub struct ListContentParams {
```
Comment (diptanu) [inline src/api.rs:283]: ListContentFilters?

File: proto/coordinator_service.proto [hunk 6]
```diff
@@ -178,6 +178,10 @@ message GetRepositoryResponse {
 message ListContentRequest {
     string repository = 1;
     string source = 2;
+    string parent_id = 3;
+
+    bool has_labels_eq_filter = 4;
```
[the remark this answers is not in the document]
Comment (jackbackes) [reply proto/coordinator_service.proto:183]: there's a distinction between an empty labels filter (Some(HashMap::new())) and not having a labels filter (None). An empty labels filter will match content that doesn't have labels.
